<a href="https://colab.research.google.com/github/ssaffiiya/ml-tasks/blob/main/DSMA_TEST_1_KHADI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


### Data preporation

In [ ]:

import pandas as pd
import numpy as np


In [ ]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [ ]:
data = pd.read_csv('2022_assignment data - Sheet1.csv')
data

week sales.brand1 feature.brand1   display.brand1 price.brand1  \
0       1   257208,785              0  0,0008017904918  1,410858252   
1       2    271358,99   0,1295018609    0,09572100309  1,398503942   
2       3    426873,23   0,4401562588     0,2302437576  1,299585368   
3       4   286039,995              0    0,04099700958  1,360167292   
4       5  308484,0644   0,1334593283    0,08613197599   1,33415596   
..    ...          ...            ...              ...          ...   
105   106   279296,745              0    0,05579531301   1,41516357   
106   107    283222,33              0    0,07857473538  1,415751817   
107   108   283704,595              0     0,2465205609  1,414666857   
108   109   293609,215              0     0,3043745193  1,414893745   
109   110   476879,955              0     0,2805601325  1,415942106   

    assortment.brand1  tv.brand1 digital.brand1 ooh.brand1 magazine.brand1  \
0         41,52883088   515838,6              0          0        66239,47   
1         41,54084823   506481,3              0          0               0   
2         43,74957835   478553,5              0          0        28108,43   
3         42,49427428   489282,1              0          0               0   
4          41,4892341   544790,9              0          0        80113,09   
..                ...        ...            ...        ...             ...   
105        44,7025922          0         152,13   72837,47               0   
106        44,4915163   617258,3         750,14   11718,75           99499   
107       44,44554651  1050923,3        7203,64          0               0   
108       44,55310634    1107153        5028,05          0               0   
109       45,74031369   376824,1           0,39          0               0   

    sales.brand2 feature.brand2 display.brand2  price.brand2  \
0      76830,795   0,5198143293   0,6296440071  0,8402281433   
1       47814,38              0   0,1736802908  0,9538375234   
2      33844,445              0   0,1682021576   1,016731465   
3       38296,24              0              0  0,9983853072   
4        37787,5              0              0  0,9959727428   
..           ...            ...            ...           ...   
105     43021,43              0              0   1,086268854   
106     43502,66              0              0   1,086244997   
107     44564,35              0              0   1,085659543   
108     43841,03              0              0   1,085222923   
109     63182,92              0              0   1,085520874   

    assortment.brand2  tv.brand2 digital.brand2 ooh.brand2  magazine.brand2  
0         13,96379027          0              0          0                0  
1         13,93096737          0              0          0                0  
2         13,70294841   98132,25              0          0                0  
3         13,84481948          0              0          0                0  
4         13,90940889  284455,75              0          0                0  
..                ...        ...            ...        ...              ...  
105       13,59557178          0              0          0                0  
106       13,57150699          0              0          0                0  
107       13,61302771          0              0          0                0  
108       13,47204176          0              0          0                0  
109        13,9535656          0              0          0                0  

[110 rows x 19 columns]

In [ ]:
data.dtypes

week                  int64
sales.brand1         object
feature.brand1       object
display.brand1       object
price.brand1         object
assortment.brand1    object
tv.brand1            object
digital.brand1       object
ooh.brand1           object
magazine.brand1      object
sales.brand2         object
feature.brand2       object
display.brand2       object
price.brand2         object
assortment.brand2    object
tv.brand2            object
digital.brand2       object
ooh.brand2           object
magazine.brand2       int64
dtype: object

In [ ]:
data = data.apply(lambda x: x.str.replace(',', '.').astype(float) if x.dtype == 'object' else x)

data.dtypes

week                   int64
sales.brand1         float64
feature.brand1       float64
display.brand1       float64
price.brand1         float64
assortment.brand1    float64
tv.brand1            float64
digital.brand1       float64
ooh.brand1           float64
magazine.brand1      float64
sales.brand2         float64
feature.brand2       float64
display.brand2       float64
price.brand2         float64
assortment.brand2    float64
tv.brand2            float64
digital.brand2       float64
ooh.brand2           float64
magazine.brand2        int64
dtype: object

### Q1

In [ ]:
desc_stats = data.describe()

desc_stats

week   sales.brand1  feature.brand1  display.brand1  \
count  110.000000     110.000000      110.000000      110.000000   
mean    55.500000  308246.783045        0.041072        0.142617   
std     31.898276   38858.510478        0.118414        0.082711   
min      1.000000  257208.785000        0.000000        0.000802   
25%     28.250000  282545.620000        0.000000        0.086924   
50%     55.500000  298655.745000        0.000000        0.131595   
75%     82.750000  320730.055000        0.000000        0.175547   
max    110.000000  476879.955000        0.658483        0.401252   

       price.brand1  assortment.brand1     tv.brand1  digital.brand1  \
count    110.000000         110.000000  1.100000e+02      110.000000   
mean       1.349152          44.828447  3.781374e+05     4816.267182   
std        0.040457           2.574797  2.925421e+05     8685.430269   
min        1.273880          38.305761  0.000000e+00        0.000000   
25%        1.310791          43.044256  1.965432e+05      324.727500   
50%        1.350414          44.802237  3.339525e+05     1575.005000   
75%        1.364957          46.759077  5.058597e+05     4968.122500   
max        1.423048          49.754567  1.191093e+06    50022.470000   

          ooh.brand1  magazine.brand1  sales.brand2  feature.brand2  \
count     110.000000       110.000000    110.000000      110.000000   
mean    82105.664000     13700.757727  44901.840420        0.043741   
std    155123.699745     26715.647402   8999.532273        0.147232   
min         0.000000         0.000000  33366.080000        0.000000   
25%         0.000000         0.000000  40260.705000        0.000000   
50%      8883.015000         0.000000  42567.655000        0.000000   
75%     85862.112500     12120.000000  46335.722500        0.000000   
max    695901.450000    113249.000000  84319.645000        0.554914   

       display.brand2  price.brand2  assortment.brand2      tv.brand2  \
count      110.000000    110.000000         110.000000     110.000000   
mean         0.045057      1.044643          12.959783   58189.560364   
std          0.125660      0.056823           0.815771  106680.115992   
min          0.000000      0.835098          11.644427       0.000000   
25%          0.000000      1.019054          12.217111       0.000000   
50%          0.000000      1.046526          12.658095       0.000000   
75%          0.002344      1.088579          13.770042   71920.875000   
max          0.650349      1.126957          14.402766  435437.750000   

       digital.brand2     ooh.brand2  magazine.brand2  
count      110.000000     110.000000       110.000000  
mean       228.863000    7687.427182       781.818182  
std        941.516614   34667.134716      8199.778268  
min          0.000000       0.000000         0.000000  
25%          0.000000       0.000000         0.000000  
50%          0.000000       0.000000         0.000000  
75%          0.000000       0.000000         0.000000  
max       7481.380000  200664.030000     86000.000000

In [ ]:
total_sales_brand1 = data['sales.brand1'].sum()
total_sales_brand2 = data['sales.brand2'].sum()

In [ ]:
# market leader
sales_mean = data[['sales.brand1', 'sales.brand2']].mean()
market_leader = 'Brand 1 (Coca-Cola)' if sales_mean['sales.brand1'] > sales_mean['sales.brand2'] else 'Brand 2 (Pepsi)'
print("Market Leader: ", market_leader)

Market Leader:  Brand 1 (Coca-Cola)


In [ ]:
total_sales_brand1,total_sales_brand2

(33907146.135, 4939202.446199999)

In [ ]:
price_comparison = data[['price.brand1', 'price.brand2']].mean()
assortment_comparison = data[['assortment.brand1', 'assortment.brand2']].mean()

print("Price Comparison:\n", price_comparison)
print()
print("Assortment Comparison:\n", assortment_comparison)


Price Comparison:
 price.brand1    1.349152
price.brand2    1.044643
dtype: float64

Assortment Comparison:
 assortment.brand1    44.828447
assortment.brand2    12.959783
dtype: float64


In [ ]:
price_comparison = data[['price.brand1', 'price.brand2']].mean()
assortment_comparison = data[['assortment.brand1', 'assortment.brand2']].mean()

print("Price Comparison:\n", price_comparison)
print()
print("Assortment Comparison:\n", assortment_comparison)


Price Comparison:
 price.brand1    1.349152
price.brand2    1.044643
dtype: float64

Assortment Comparison:
 assortment.brand1    44.828447
assortment.brand2    12.959783
dtype: float64


In [ ]:
ad_spending_coca_cola = data[['tv.brand1', 'ooh.brand1', 'magazine.brand1', 'digital.brand1']].sum()
ad_spending_pepsi = data[['tv.brand2', 'ooh.brand2', 'magazine.brand2', 'digital.brand2']].sum()

print(f"Advertising spending for Coca-Cola (TV, OOH, Magazine, Digital): \n{ad_spending_coca_cola}")
print(f"Advertising spending for Pepsi (TV, OOH, Magazine, Digital): \n{ad_spending_pepsi}")


avg_promo_coca_cola = promo_spending_coca_cola.mean()
avg_promo_pepsi = promo_spending_pepsi.mean()

print(f"Average promo spending for Coca-Cola: {avg_promo_coca_cola}")
print(f"Average promo spending for Pepsi: {avg_promo_pepsi}")

Advertising spending for Coca-Cola (TV, OOH, Magazine, Digital): 
tv.brand1          41595116.16
ooh.brand1          9031623.04
magazine.brand1     1507083.35
digital.brand1       529789.39
dtype: float64
Advertising spending for Pepsi (TV, OOH, Magazine, Digital): 
tv.brand2          6400851.64
ooh.brand2          845616.99
magazine.brand2      86000.00
digital.brand2       25174.93
dtype: float64
Average promo spending for Coca-Cola: 10.102917710255902
Average promo spending for Pepsi: 4.88386264213115


In [ ]:
promo_spending_coca_cola = data[['feature.brand1', 'display.brand1']].sum()
promo_spending_pepsi = data[['feature.brand2', 'display.brand2']].sum()

print(f"Promo spending for Coca-Cola (feature + display): \n{promo_spending_coca_cola}")
print(f"Promo spending for Pepsi (feature + display): \n{promo_spending_pepsi}")


Promo spending for Coca-Cola (feature + display): 
feature.brand1     4.517936
display.brand1    15.687900
dtype: float64
Promo spending for Pepsi (feature + display): 
feature.brand2    4.811461
display.brand2    4.956265
dtype: float64


Coca-Cola (Brand 1) is the market leader with total sales of approximately 33.9 million, compared to Pepsi's 4.9 million, indicating its dominance. Coca-Cola's average price is higher at 1.35, while Pepsi's is 1.04, suggesting Coca-Cola is positioned as a more premium product. Coca-Cola also offers a broader assortment with an average of 44.83, compared to Pepsi's 12.96, showing its wider product variety.

In terms of promotional spending, Coca-Cola significantly outspends Pepsi. Coca-Cola's total promotional spend (feature + display) is about 20.2 million, compared to Pepsi's 9.8 million. Additionally, Coca-Cola spends more across all advertising instruments—TV, OOH, Magazine, and Digital—totaling 54.7 million, while Pepsi spends only 7.2 million. This shows Coca-Cola’s more aggressive and diversified advertising strategy.

Overall, Coca-Cola leads in sales, pricing, product assortment, and marketing spend. Pepsi, with its more competitive pricing, spends much less on marketing, which could affect its market share growth.

### Q2

In [ ]:
data_transformed = pd.DataFrame()
data_transformed['log_sales_brand2'] = np.log(data['sales.brand2'])
data_transformed['log_feature_brand2'] = np.log(data['feature.brand2'] + 1)
data_transformed['log_display_brand2'] = np.log(data['display.brand2'] + 1)
data_transformed['log_assortment_brand2'] = np.log(data['assortment.brand2'])
data_transformed['log_price_brand2'] = np.log(data['price.brand2'])
data_transformed['log_tv_brand2'] = np.log(data['tv.brand2'] + 1)
data_transformed['log_digital_brand2'] = np.log(data['digital.brand2'] + 1)
data_transformed['log_ooh_brand2'] = np.log(data['ooh.brand2'] + 1)
data_transformed['log_magazine_brand2'] = np.log(data['magazine.brand2'] + 1)


In [ ]:
X = data_transformed[[
    'log_feature_brand2',
    'log_display_brand2',
    'log_assortment_brand2',
    'log_price_brand2',
    'log_tv_brand2',
    'log_digital_brand2',
    'log_ooh_brand2',
    'log_magazine_brand2',
]]

In [ ]:
X = sm.add_constant(X)

In [ ]:
model = sm.OLS(data_transformed['log_sales_brand2'], X).fit()


In [ ]:
model_summary = model.summary()
model_summary


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:       log_sales_brand2   R-squared:                       0.641
Model:                            OLS   Adj. R-squared:                  0.612
Method:                 Least Squares   F-statistic:                     22.51
Date:                Mon, 23 Dec 2024   Prob (F-statistic):           2.33e-19
Time:                        15:43:51   Log-Likelihood:                 94.029
No. Observations:                 110   AIC:                            -170.1
Df Residuals:                     101   BIC:                            -145.8
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                    10.0030      0.484     20.680      0.000       9.043      10.963
log_feature_brand2        1.0239      0.198      5.168      0.000       0.631       1.417
log_display_brand2       -0.0634      0.231     -0.274      0.784      -0.522       0.395
log_assortment_brand2     0.2641      0.187      1.416      0.160      -0.106       0.634
log_price_brand2         -0.4752      0.284     -1.673      0.097      -1.039       0.088
log_tv_brand2            -0.0006      0.002     -0.285      0.777      -0.005       0.003
log_digital_brand2       -0.0009      0.005     -0.184      0.854      -0.010       0.009
log_ooh_brand2            0.0053      0.003      1.554      0.123      -0.001       0.012
log_magazine_brand2       0.0099      0.010      1.027      0.307      -0.009       0.029
==============================================================================
Omnibus:                       16.636   Durbin-Watson:                   1.016
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               25.870
Skew:                           0.696   Prob(JB):                     2.41e-06
Kurtosis:                       4.925   Cond. No.                         368.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Based on the regression results, the model explains approximately 64.1% of the variation in Pepsi's sales (ùëÖ2=0.641). This indicates a moderately strong relationship between the marketing mix variables and sales.
Among the predictors, feature advertising is statistically significant (𝑝<0.01) and has a large positive coefficient (β=1.024). This implies that a 1% increase in feature advertising is associated with an approximately 1.02% increase in Pepsi’s sales. Clearly, this variable is a major driver of sales growth.
Conversely, other variables such as display advertising, price, TV, and digital advertising are not statistically significant at the 5% level. Although the coefficient for price (β=−0.475) suggests a negative relationship with sales, the p-value of 0.097 indicates weaker evidence against the null hypothesis. Thus, price changes may have some effect, but this result should be interpreted cautiously.
In summary, the findings highlight the critical importance of feature advertising for Pepsi's sales performance, while other marketing mix elements show limited statistical evidence of impact in this model. Further investigation could explore interaction effects or nonlinear relationships.

### Q3

In [ ]:
def adstock(series, decay):
    adstocked = np.zeros(len(series))
    adstocked[0] = series.iloc[0]
    for t in range(1, len(series)):
        adstocked[t] = series.iloc[t] + decay * adstocked[t-1]
    return adstocked

In [ ]:
lambda_decay = 0.6
data['adstock_tv'] = adstock(data['tv.brand2'], lambda_decay)
data['adstock_ooh'] = adstock(data['ooh.brand2'], lambda_decay)
data['adstock_magazine'] = adstock(data['magazine.brand2'], lambda_decay)
data['adstock_digital'] = adstock(data['digital.brand2'], lambda_decay)


In [ ]:
data_transformed = pd.DataFrame()
data_transformed['log_sales_brand2'] = np.log(data['sales.brand2'])
data_transformed['log_feature_brand2'] = np.log(data['feature.brand2'] + 1)
data_transformed['log_display_brand2'] = np.log(data['display.brand2'] + 1)
data_transformed['log_assortment_brand2'] = np.log(data['assortment.brand2'])
data_transformed['log_price_brand2'] = np.log(data['price.brand2'])
data_transformed['log_adstock_tv'] = np.log(data['adstock_tv'] + 1)
data_transformed['log_adstock_ooh'] = np.log(data['adstock_ooh'] + 1)
data_transformed['log_adstock_magazine'] = np.log(data['adstock_magazine'] + 1)
data_transformed['log_adstock_digital'] = np.log(data['adstock_digital'] + 1)


In [ ]:
X = data_transformed[[
    'log_feature_brand2',
    'log_display_brand2',
    'log_assortment_brand2',
    'log_price_brand2',
    'log_adstock_tv',
    'log_adstock_ooh',
    'log_adstock_magazine',
    'log_adstock_digital'
]]

In [ ]:
X = sm.add_constant(X)


In [ ]:
model = sm.OLS(data_transformed['log_sales_brand2'], X).fit()


In [ ]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:       log_sales_brand2   R-squared:                       0.708
Model:                            OLS   Adj. R-squared:                  0.685
Method:                 Least Squares   F-statistic:                     30.60
Date:                Mon, 23 Dec 2024   Prob (F-statistic):           8.87e-24
Time:                        16:03:21   Log-Likelihood:                 105.43
No. Observations:                 110   AIC:                            -192.9
Df Residuals:                     101   BIC:                            -168.6
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                     8.8599      0.550     16.109      0.000       7.769       9.951
log_feature_brand2        0.9425      0.180      5.228      0.000       0.585       1.300
log_display_brand2       -0.2037      0.211     -0.966      0.336      -0.622       0.215
log_assortment_brand2     0.7122      0.210      3.390      0.001       0.295       1.129
log_price_brand2         -1.1203      0.291     -3.856      0.000      -1.697      -0.544
log_adstock_tv           -0.0026      0.004     -0.587      0.559      -0.011       0.006
log_adstock_ooh           0.0124      0.003      4.188      0.000       0.007       0.018
log_adstock_magazine      0.0114      0.004      2.894      0.005       0.004       0.019
log_adstock_digital      -0.0048      0.004     -1.086      0.280      -0.014       0.004
==============================================================================
Omnibus:                       21.971   Durbin-Watson:                   1.089
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               35.429
Skew:                           0.898   Prob(JB):                     2.03e-08
Kurtosis:                       5.123   Cond. No.                         826.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Based on the regression results, the model with adstock transformation explains approximately 70.8% of the variation in Pepsi's sales (R 2=0.708). This indicates a strong relationship between the marketing mix variables and sales when accounting for the longer-term effects of advertising.
Among the adstocked advertising variables, OOH  advertising (β=0.0124, p<0.01) and magazine advertising (β=0.0114, p<0.01) have statistically significant positive effects on sales. This suggests that a 1% increase in OOH or magazine adstock leads to a 0.012% and 0.011% increase in sales, respectively. These findings highlight the potential longer-term impact of these advertising channels on sales.
In contrast, TV (p=0.559) and digital (p=0.280) advertising adstock effects are not statistically significant, indicating limited evidence of their long-term impact on sales in this dataset.
For other marketing mix elements, feature advertising (β=0.9425, p<0.01) and assortment (β=0.7122, p<0.01) remain significant drivers of sales, while price (β=−1.1203, p<0.01) has a substantial negative effect.
In summary, the findings suggest that feature ads, assortment, and price play critical roles in driving sales, while OOH and magazine advertising exhibit significant longer-term effects. TV and digital advertising show limited evidence of impact in this model. Further exploration of interaction effects or alternate decay rates might provide additional insights.

### Q4


Yes, the model likely omits important variables that could impact sales. For instance, factors like competitor advertising (e.g., Coca-Cola’s efforts), seasonality (e.g., holiday peaks), and economic conditions (e.g., inflation) are not included. These could significantly influence Pepsi’s sales.
Excluding such variables risks omitted variable bias, where coefficients for included variables may capture effects of omitted ones. For example, if feature advertising often aligns with seasonal trends, its effect could be overstated without controlling for seasonality. Similarly, competitor actions may dilute Pepsi’s marketing impact, making some estimates less reliable.
Other missing factors might include store promotions, distribution coverage, and brand loyalty, which are critical sales drivers. Without these, the model may not fully explain sales variations, reducing its explanatory power.
In summary, while the model provides useful insights, omitted variables could distort results, highlighting the need for a more comprehensive analysis.